pkg/syncutil/cachemanager/cachemanager.go

```diff
@@ -41,17 +186,19 @@ func (c *CacheManager) AddObject(ctx context.Context, instance *unstructured.Uns
 	}
 
 	syncKey := syncutil.GetKeyForSyncMetrics(instance.GetNamespace(), instance.GetName())
-	_, err = c.opa.AddData(ctx, instance)
-	if err != nil {
-		c.syncMetricsCache.AddObject(
-			syncKey,
-			syncutil.Tags{
-				Kind:   instance.GetKind(),
-				Status: metrics.ErrorStatus,
-			},
-		)
+	if c.watchedSet.Contains(gvk) {
+		_, err = c.opa.AddData(ctx, instance)
+		if err != nil {
+			c.syncMetricsCache.AddObject(
+				syncKey,
+				syncutil.Tags{
+					Kind:   instance.GetKind(),
+					Status: metrics.ErrorStatus,
+				},
+			)
 
-		return err
+			return err
+		}
 	}
 
 	c.tracker.ForData(instance.GroupVersionKind()).Observe(instance)
```

Comment:
Let's look at the old code:

addition for not namespace excluded
https://github.com/open-policy-agent/gatekeeper/blob/605dd6339c174654002b27bd6bc57df624879c79/pkg/controller/sync/sync_controller.go#L217-L242

addition for namespace excluded
https://github.com/open-policy-agent/gatekeeper/blob/605dd6339c174654002b27bd6bc57df624879c79/pkg/controller/sync/sync_controller.go#L196-L201

When it comes to tracking and metrics cache, I think this behavior is consistent. Recall that the opa client being used by the sync_c in the example above is the filtered one, equivalent to our `if c.watchedSet.Contains(gvk)` check.

IIUC, we cancel expectations for what is namespace excluded for cleanup and observe in the happy path bc we technically have "observed" the gvk's instance.

Reply:
interesting, I didn't see issues  w the tracker around this when testing with multiple sources sync that synced the same gvk. Do you have a fail scenario off the top of your head here that comes fast?

> Specifically, t.config.Populated() will no longer be sufficient.

there's no docstrings on `t.config.Populated()` so I am trying to parse the statement above. Please correct me if I'm wrong, but `Populated()` means that we stopped looking for `Expectations` which is set by `ExpectationsDone()` which is meant to be called from `trackConfig` for when we have a `Config` resource at startup. In that case, we want to wait on syncing all the instances for the gvks in `syncOnly`, right ?

In adding another source type, `syncset`, (type not name), we will likely need to add another `object tracker?  like `t.config` to make sure the ready tracker functions well for when we want to wait on startup for all `syncsets` expectations to be met. Or we could stuff all `syncset` GVKs' expectation into the `t.config` for the beginning.

It seems the only time we may run into an issue is on startup with multiple sync types. And that issue would be not tracking all expectations properly, is that correct?

I'm trying to understand if this is an issue for this PR or for when we add the syncset controller (after this PR)? Is this blocking this PR right now?

Reply:
Not blocking the PR, will cause readiness to break if it's not addressed by the time SyncSet controller is added.

"Populated" means that the initial listing for that resource has been completed, meaning we expect no further expectations to be established from that config source.

This lets us:

* Cache observations for any objects that are observed before expectations are fully populated, to avoid expectations never being satisfied b/c we observed the object that satisfied it too early.

* Avoid continuing to cache observations speculatively once expectations are established, to avoid unnecessary memory usage.

Reply:
And, yeah ExpectationsDone() is how we signal to a tracker that there are no more inbound expectations, so Populated() is now true.